### ISSUE TITLE
array hashing

### ISSUE THREAD

**Issue by michelangelo21** [JuliaLang/julia#41880]
Recently I discovered that hashing higher-dimensional arrays is faster when flatting them first.

```julia
using BenchmarkTools

@benchmark hash($rand(512)) # 8.072 μs
@benchmark hash($rand(8,8,8)) # 26.493 μs
@benchmark hash(vec($rand(8,8,8))) # 8.239 μs 
```

shouldn't arrays of primitive types have optimized methods for hash?

**Comment by KristofferC** [comment]
Looks fixed in Julia 1.7. You can confirm by downloading the beta.

**Comment by michelangelo21** [comment]
For size 8^3 difference is negligible (2.290, 2.547, 2.341  μs), but for 100^3, still the second line is a bit slower 
```julia
using BenchmarkTools

@benchmark hash($rand(100^3))  # 1.738 ms
@benchmark hash($rand(100,100,100))  # 2.761 ms
@benchmark hash(vec($rand(100,100,100)))  # 1.780 ms
```
```julia
julia> versioninfo()
Julia Version 1.7.0-beta3.0
Commit e76c9dad42 (2021-07-07 08:12 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Core(TM) i7-8565U CPU @ 1.80GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-12.0.0 (ORCJIT, skylake)
```

**Comment by giordano** [comment]
The benchmarks above are basically measuring the time to draw random numbers to fill the arrays: `$rand(...)` is interpolating the `rand` symbol, not the value of the `rand` call, you need more parentheses to achieve that.  The fact that `@benchmark` showed memory allocations (not reported above) should have been a big giveaway, `hash` shouldn't allocate any memory.

That said, when doing the benchmark the right way, one can see the difference is real and is still there on `master` (dc344285d5):
```julia-repl
julia> using BenchmarkTools

julia> @btime hash($(rand(100 ^ 3)));
  163.125 μs (0 allocations: 0 bytes)

julia> @btime hash($(rand(100, 100, 100)));
  585.541 μs (0 allocations: 0 bytes)

julia> @btime hash($(vec(rand(100, 100, 100)))); # this is equivalent to the first one, as the timing shows as well
  163.125 μs (0 allocations: 0 bytes)
```

**Comment by nsajko** [comment]
Why close? Seems like it should be possible to speed up hashing of multidimensional arrays.

**Comment by mbauman** [comment]
Yes, we could change hashing to hash pairs using the linear index `linidx => elt` here:

https://github.com/JuliaLang/julia/blob/5835c3b69e4e0f47eeb9a512d91622b50ad3423c/base/abstractarray.jl#L3628

We could then have a fastpath for `IndexLinear()` arrays that just uses linear indexing, but that additionally requires a `findprev` that works on linear indices (instead of `keys`).  That would address most of the time difference, all coming from this line:

https://github.com/JuliaLang/julia/blob/5835c3b69e4e0f47eeb9a512d91622b50ad3423c/base/abstractarray.jl#L3647

In fact, more work could surely be done to optimize the very naive and generic `findprev` — for both linear and cartesian Indices.

https://github.com/JuliaLang/julia/blob/5835c3b69e4e0f47eeb9a512d91622b50ad3423c/base/array.jl#L2579-L2590
